lib/Module.js

```diff
@@ -171,18 +164,40 @@ class Module extends DependenciesBlock {
 		return false;
 	}
 
+
+	// Remove oldChunk from and add newChunks to the module's reasons.
+	// Use a hash map to avoid adding duplicate chunks. This approach is used
+	// rather than indexOf for performance reasons.
 	rewriteChunkInReasons(oldChunk, newChunks) {
-		this.reasons.forEach(r => {
+		// create the newChunksMap outside the reasons loop to avoid duplicating
+		// the work.
+		const newChunkMap = {};
+		newChunks.forEach(chunk => {
+			newChunkMap[chunk.id] = chunk;
+		});
+
+		for(let reasonIndex = 0; reasonIndex < this.reasons.length; reasonIndex++) {
+			const r = this.reasons[reasonIndex];
 			if(!r.chunks) {
 				if(!r.module._chunks.has(oldChunk))
 					return;
 				r.chunks = Array.from(r.module._chunks);
 			}
-			r.chunks = r.chunks.reduce((arr, c) => {
-				addToSet(arr, c !== oldChunk ? [c] : newChunks);
-				return arr;
+			const chunksMap = {};
+			r.chunks.forEach(chunk => {
+				if(chunk !== oldChunk) {
+					chunksMap[chunk.id] = chunk;
+				}
+			});
+			// override any matching chunkIds
+			Object.assign(chunksMap, newChunkMap);
+
+			// Convert each entry in the chunksMap into a value in an array.
+			r.chunks = Object.keys(chunksMap).reduce((acc, key) => {
```

Comment:
you could definitely create a PR with that wording and see where it goes. Or maybe ping @TheLarkInn or @bebraw .